### DISCUSSION TITLE
More detailed step-by-step guide to getting started

### DISCUSSION THREAD

**Discussion by ajkessel** [marclove/pythreads#12] [Q&A] [answered]
The readme seems to be missing some details about how you go from creating an app in the developer console to actually getting a working prototype registered and functional. Am I missing an example of a minimum viable script somewhere? Just looking for a model that goes through the initial authentication workflow from python.

**Comment by ajkessel** [comment]
Through some trial and error, I figured out a minimal shell script to get an access token:
```
curl --head -qX GET "https://threads.net/oauth/authorize?client_id=$THREADS_APP_ID&redirect_uri=$THREADS_REDIRECT_URI&scope=$SCOPE&response_type=code&state=$STATE" | grep -i location | grep -oi 'https://www.threads.net.*'
echo -n enter code:
read CODE
token=$(curl -sX POST "https://graph.threads.net/oauth/access_token" -d "client_id=$THREADS_APP_ID&client_secret=$THREADS_API_SECRET&code=$CODE&grant_type=authorization_code&redirect_uri=$THREADS_REDIRECT_URI" |jq -r '.access_token')
echo got token $token
```
Now I'm wondering if there is any way to retrieve Media ID from a Threads URL. The documentation is pretty sparse on retrieving content as opposed to posting it.

**Comment by marclove** [comment]
Hi @ajkessel, did the instructions in the [Authentication & Authorization section](https://github.com/marclove/pythreads?tab=readme-ov-file#authentication--authorization) not work for you?

As for documentation, the methods follow the [Threads API docs](https://developers.facebook.com/docs/threads) fairly closely. Have you reviewed those for what you're trying to do?

**Reply by ajkessel** [reply]
I was looking for the minimal code necessary for the authentication workflow. The instructions in the readme seem to presuppose some knowledge that I'm apparently missing--how are you setting things up for Threads.authorization_url() to function? Is that code that comes with this module or that the developer needs to create? Ideally, there would be a working python code example that takes the app ID and secret and comes back with a token. My simple shell script basically does that. 

In terms of the API, maybe I'm looking for functionality that doesn't exist. If all you have is a public Threads URL and want to inspect the object there, can you do anything with that via the official API? Or does the API presuppose you already have a media ID for any item you want to inspect?

**Reply by marclove** [reply, marked as answer]
The readme should be the minimal code necessary for the authentication flow. If you run into problems, let me know. Once you set up the environment variables, you should be able to complete authentication using the exact steps outlined in the readme.

This library is meant to be used in the context of a web app of some kind since that is the OAuth flow that the API supports, so you will need a redirect url that you've configured your app with on Meta's developer console that matches the THREADS_REDIRECT_URI env var. Once you have that set up, the steps outlined in the readme should automatically handle the results of that redirect to turn the auth code into a long-lived token.

As for retrieving media, I'd suggest reading the relevant part of the [API docs](https://developers.facebook.com/docs/threads/threads-media) to understand the endpoints involved for your particular use case.

**Reply by marclove** [reply]
Also, in addition to the API docs, I'd recommend looking at the test suite for examples on how to use this wrapper specifically.

**Reply by ajkessel** [reply]
Maybe I'm dense here, but I don't see how you execute Threads.authorization_url(), which is the first step in your instructions -- what provides that function?

I have reviewed the API docs carefully and searched for examples, so maybe the functionality just doesn't exist--if you receive a public Threads URL, there is no way to interact with it via the API?

**Comment by ajkessel** [comment]
OK, I made some progress. Here is the step-by-step I suggest:
```
# openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -sha256 -days 3650 -nodes -subj "/C=XX/ST=StateName/L=CityName/O=CompanyName/OU=CompanySectionName/CN=CommonNameOrHostname"
# export THREADS_SSL_KEY_FILEPATH=`realpath ./key.pem`
# export THREADS_SSL_CERT_FILEPATH=`realpath` ./cert.pem`
# export THREADS_APP_ID=(Threads App ID)
# export THREADS_API_SECRET=(Threads App API Secret)
# export THREADS_REDIRECT_URI=(URL of page you have defined in app that can display the code)
```
Then, in Python:
```
from pythreads import threads
auth_url, state_key = threads.Threads.authorization_url()
```
The point I missed was a simple one -- you need to import threads from pythreads and then refer to the function threads.Threads.authorization_url().

I can suggest a full little Python script to pull this altogether -- for the user who is just getting their arms around the API and wants to be able to experiment with it from the command-line.
